aidial_assistant/json_stream/json_parser.py

```diff
@@ -38,34 +46,43 @@ def string_node(node: JsonNode) -> JsonString:
     return node
 
 
-class JsonParser:
-    @staticmethod
-    @asynccontextmanager
-    async def parse(stream: CharacterStream) -> AsyncGenerator[JsonNode, Any]:
-        root = JsonRoot()
+class RootNodeResolver(NodeResolver):
+    @override
+    async def resolve(self, stream: CharacterStream) -> JsonNode:
         try:
-            async with TaskGroup() as tg:
-                task = tg.create_task(JsonParser._parse_root(root, stream))
-                try:
-                    yield await root.node()
-                finally:
-                    await task
-        except ExceptionGroup as e:
-            raise e.exceptions[0]
+            normalised_stream = JsonNormalizer(stream)
+            char = await normalised_stream.apeek()
+            if char == JsonObject.token():
```

Comment:
Maybe introduce the same interface for all these classes, so that dispatching looks uniform:

```python
class RootNodeResolver(NodeResolver):
    @override
    async def resolve(self, stream: CharacterStream) -> JsonNode:
        try:
            normalised_stream = JsonNormalizer(stream)
            char = await normalised_stream.apeek()
            if JsonObject.is_token():
                return JsonObject.parse(stream, self)

            if JsonArray.is_token():
                return JsonArray.parse(stream, self)

            if JsonString.is_token():
                return JsonString.parse(stream)

            if JsonNumber.is_token(char):
                return JsonNumber.parse(stream)

            if JsonNull.is_token(char):
                return JsonNull.parse(stream)

            if JsonBoolean.is_token(char):
                return JsonBoolean.parse(stream)
        except StopAsyncIteration:
            raise unexpected_end_of_stream_error(stream.char_position)

        raise unexpected_symbol_error(char, stream.char_position)
```